tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs

```diff
@@ -0,0 +1,290 @@
+                {
+                    await Task.Delay(_pollInterval, _cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    // We are shutting down, so don't do anything about it
+                }
+            }
+        }
+
+        public void RegisterProduct(Product product)
+        {
+            _products.TryAdd(product.Name, product);
+        }
+
+        public void UnregisterProduct(string productName)
+        {
+            _products.TryRemove(productName, out _);
+        }
+
+        private async Task Poll()
+        {
+            await SemaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                var request = BuildRequest();
+                var response = _remoteConfigurationApi.GetConfigs(request).Result;
+
+                if (response?.Targets?.Signed != null)
+                {
+                    ProcessResponse(response);
+                    _targetsVersion = response.Targets.Signed.Version;
+                }
+            }
+            catch (Exception e)
+            {
+                _lastPollError = e.Message;
+            }
+            finally
+            {
+                SemaphoreSlim.Release();
+            }
+        }
+
+        private GetRcmRequest BuildRequest()
+        {
+            var appliedConfigurations = _products.Values.SelectMany(pair => pair.AppliedConfigurations.Values).ToList();
+            var cachedTargetFiles = appliedConfigurations.Select(cache => new RcmCachedTargetFile(cache.Path.Path, cache.Length, cache.Hashes.Select(kp => new RcmCachedTargetFileHash(kp.Key, kp.Value)).ToList())).ToList();
+
+            return new GetRcmRequest(BuildRequestClient(), cachedTargetFiles);
+
+            RcmClient BuildRequestClient()
+            {
+                return new RcmClient(_id, _products.Keys.ToList(), BuildRcmClientTracer(), BuildRequestClientState());
+            }
+
+            RcmClientTracer BuildRcmClientTracer()
+            {
+                return new RcmClientTracer(_runtimeId, _tracerVersion, _service, _env, _appVersion);
+            }
+
+            RcmClientState BuildRequestClientState()
+            {
+                var configStates = appliedConfigurations.Select(cache => new RcmConfigState(cache.Path.Id, cache.Version, cache.Path.Product)).ToList();
+                return new RcmClientState(_rootVersion, _targetsVersion, configStates, _lastPollError != null, _lastPollError);
+            }
+        }
+
+        private void ProcessResponse(GetRcmResponse response)
+        {
+            var actualConfigPath =
+                response.ClientConfigs
+                        .Select(RemoteConfigurationPath.FromPath)
+                        .ToDictionary(path => path.Path);
+
+            var changedConfigurationsByProduct = GetChangedConfigurations()
+               .GroupBy(config => config.Path.Product);
```

Comment:
Why grouping it here if you could directly return the structure you need?

Reply:
For readability of the code.